<launch>

  <!-- important generally, but specifically utilised by the current app manager -->
  <param name="robot/name" value="$(optenv ROBOT turtlebot)"/>
  <param name="robot/type" value="turtlebot"/>


  <param name="robot_description" command="cat $(find gibson-ros)/turtlebot/turtlebot.urdf"/>

  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
  </node>

  <node name="turtlebot_gibson_sim" pkg="gibson-ros" type="turtlebot_rgbd.py" output="screen">
  </node>

    <node name="gibson_clock" pkg="gibson-ros" type="simulation_clock.py">
  </node>

  <node name="gibson_sim_noise" pkg="gibson-ros" type="noise_injection.py">
  </node>

  <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="5.0" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find gibson-ros)/rviz/navigation.rviz" />

  <!-- Velocity muxer -->
  <node pkg="nodelet" type="nodelet" name="mobile_base_nodelet_manager" args="manager"/>
  <node pkg="nodelet" type="nodelet" name="cmd_vel_mux"
        args="load yocs_cmd_vel_mux/CmdVelMuxNodelet mobile_base_nodelet_manager">
    <param name="yaml_cfg_file" value="$(find turtlebot_bringup)/param/mux.yaml" />
    <remap from="cmd_vel_mux/output" to="mobile_base/commands/velocity_raw"/>
    <remap from="cmd_vel_mux/input/navi" to="/navigation_velocity_smoother/raw_cmd_vel" />
  </node>

 <include file="$(find turtlebot_teleop)/launch/ps3_teleop.launch">
 </include>

  <node pkg="depthimage_to_laserscan" type="depthimage_to_laserscan" name="depthimage_to_laserscan"  output="screen">
    <param name="scan_height" value="2"/>
    <param name="output_frame_id" value="/camera_depth_frame"/>
    <param name="range_min" value="0.45"/>
    <remap from="image" to="/gibson_ros/camera/depth/image"/>
    <remap from="scan" to="/scan"/>
  </node>

  <arg name="map_file" default="$(find gibson-ros)/map/gates.yaml"/>
 <node pkg="map_server" type="map_server"  name="map_server" args="$(arg map_file)" />
 <node pkg="amcl" type="amcl"  name="amcl">
   <param name="min_particles" value="500"/>
  <param name="kld_err" value="0.05"/>
  <param name="initial_cov_aa" value="1.57"/>
 </node>


  <arg name="odom_frame_id"   default="odom"/>
  <arg name="base_frame_id"   default="base_footprint"/>
  <arg name="global_frame_id" default="map"/>
  <arg name="odom_topic" default="odom" />
  <arg name="laser_topic" default="scan" />
  <arg name="custom_param_file" default="$(find gibson-ros)/param/dummy.yaml"/>

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam file="$(find gibson-ros)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find gibson-ros)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find gibson-ros)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find gibson-ros)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find gibson-ros)/param/dwa_local_planner_params.yaml" command="load" />
    <rosparam file="$(find gibson-ros)/param/move_base_params.yaml" command="load" />
    <rosparam file="$(find gibson-ros)/param/global_planner_params.yaml" command="load" />
    <rosparam file="$(find gibson-ros)/param/navfn_global_planner_params.yaml" command="load" />
    <!-- external params file that could be loaded into the move_base namespace -->
    <rosparam file="$(arg custom_param_file)" command="load" />

    <!-- reset frame_id parameters using user input data -->
    <param name="global_costmap/global_frame" value="$(arg global_frame_id)"/>
    <param name="global_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="local_costmap/global_frame" value="$(arg odom_frame_id)"/>
    <param name="local_costmap/robot_base_frame" value="$(arg base_frame_id)"/>
    <param name="DWAPlannerROS/global_frame_id" value="$(arg odom_frame_id)"/>

    <remap from="cmd_vel" to="navigation_velocity_smoother/raw_cmd_vel"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <remap from="scan" to="$(arg laser_topic)"/>
  </node>



</launch>